File: src/python/algorithms/interval/6.meetingRooms3/meeting_rooms3.py
"""
Meeting Rooms III

Given n rooms and a list of meetings, return the room that held the most meetings.
If there are ties, return the room with the lowest number.

Time Complexity: O(m * log n) where m is number of meetings
Space Complexity: O(n)
"""

import heapq

def most_booked(n, meetings):
    """
    Using two heaps - available rooms and busy rooms.
    TC:O(m * log n) + O(m * log m) => O(m * log m), SC: O(n)
    """
    # Sort meetings by start time
    meetings.sort()
    
    # min-heap of available rooms (room numbers)
    available = list(range(n))
    heapq.heapify(available)
    
    # min-heap of occupied rooms: (end_time, room)
    occupied = []
    
    # Count how many times each room is used
    count = [0] * n
    
    for start, end in meetings:
        duration = end - start
        
        # 1. Free up rooms that have finished
        while occupied and occupied[0][0] <= start:
            end_time, room = heapq.heappop(occupied)
            heapq.heappush(available, room)
        
        # 2. If a room is available → assign it
        if available:
            room = heapq.heappop(available)
            heapq.heappush(occupied, (end, room))
        
        else:
            # 3. No room available → delay meeting
            end_time, room = heapq.heappop(occupied)
            new_end = end_time + duration
            heapq.heappush(occupied, (new_end, room))
        
        count[room] += 1
    
    # Return room with maximum bookings (smallest index if tie)
    return count.index(max(count))

# Test cases
if __name__ == "__main__":
    test_cases = [
        {"n": 2, "meetings": [[0, 10], [1, 5], [2, 7], [3, 4]]},  # 0
        {"n": 3, "meetings": [[1, 20], [2, 10], [3, 5], [4, 9], [6, 8]]},  # 1
    ]
    
    for test in test_cases:
        print(f"Rooms: {test['n']}")
        print(f"Meetings: {test['meetings']}")
        print(f"Most booked room: {most_booked(test['n'], [m[:] for m in test['meetings']])}")
        print()

File: src/java1/algorithms/interval/meetingRooms3/meetingRooms3.java
package java1.algorithms.interval.meetingRooms3;

import java.util.Arrays;
import java.util.PriorityQueue;

public class meetingRooms3 {
    //Using two min heaps(freeRooms, busyRooms) TC: O(n log n) SC: O(n)
    private int mostBooked(int n, int[][] meetings) {
        //Sort by start times;
        Arrays.sort(meetings, (a, b) -> a[0]-b[0]);
        
        PriorityQueue<Integer> freeRooms = new PriorityQueue<>();
        for(int i=0; i<n; i++) {
            freeRooms.offer(i);
        }

        PriorityQueue<long[]> busyRooms = new PriorityQueue<>((a, b) -> {
            if(a[0] == b[0]) return Long.compare(a[1], b[1]);

            return Long.compare(a[0], b[0]);
        });

        int[] count = new int[n];

        for(int[] meeting: meetings) {
            int start = meeting[0];
            int end = meeting[1];
            int duration = end - start;

            //release all finished meeting rooms
            while(!busyRooms.isEmpty() && busyRooms.peek()[0] <= start) {
                freeRooms.offer((int)busyRooms.poll()[1]);
            }

            int room;
            if(!freeRooms.isEmpty()) {
                room = freeRooms.poll();
                busyRooms.offer(new long[]{end, room});
            } else {
                long[] earliest = busyRooms.poll();
                room = (int) earliest[1];
                long newEnd = earliest[0] + duration;
                busyRooms.offer(new long[]{newEnd, room});
            }
            count[room]++;
        }

        int roomWithMaxMeetings = 0;
        for(int room=1; room<n; room++) {
            if(count[room] > count[roomWithMaxMeetings]) {
                roomWithMaxMeetings = room;
            }
        }

        return roomWithMaxMeetings;
    }
}
